File: java/senscord/src/senscord_types/ObjectDetectionData.java
/*
 * SPDX-FileCopyrightText: 2020-2022 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package senscord_types;

import com.sun.jna.Structure;

import java.util.Arrays;
import java.util.List;

/**
 * Object Detection Data.
 *
 * @see "senscord::ObjectDetectionData"
 */
public class ObjectDetectionData extends Structure {

    public static class ByReference extends ObjectDetectionData implements Structure.ByReference {
    }

    public int count;
    public DetectedObjectInformation.ByReference data;

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("count", "data");
    }

    public ObjectDetectionData() {
        this.data = new DetectedObjectInformation.ByReference();
    }

    /**
     * Get data.
     *
     * @return data
     */
    public DetectedObjectInformation[] GetData() {
        DetectedObjectInformation[] retData;
        if ((count == 0) || (data == null)) {
            retData = new DetectedObjectInformation[1];
            retData[0] = new DetectedObjectInformation();
            retData[0].class_id = 0;
            retData[0].score = 0;
            retData[0].box.bottom = 0;
            retData[0].box.left = 0;
            retData[0].box.right = 0;
            retData[0].box.top = 0;
        } else {
            retData = (DetectedObjectInformation[]) data.toArray(count);
        }
        return retData;
    }

    /**
     * Get size of data.
     *
     * @return size
     */
    public int GetCount() {
        return this.count;
    }
}

File: csharp/lib/RawData/ObjectDetectionData.cs
/*
 * SPDX-FileCopyrightText: 2017-2022 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

using System.Collections.Generic;
using MessagePack;

namespace SensCord
{
    /// <summary>
    /// Object detection data.
    /// </summary>
    [MessagePackObject]
    public class ObjectDetectionData
    {
        /// <summary>
        /// Raw data type.
        /// </summary>
        [IgnoreMember]
        public string Type { get; } = RawDataType.ObjectDetection;

        /// <summary>
        /// ObjectDetection data.
        /// </summary>
        [Key("data")]
        public List<DetectedObjectInformation> Data { get; set; } =
            new List<DetectedObjectInformation>();

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return $"{this.Type}({string.Join(", ", this.Data)})";
        }
    }

    /// <summary>
    /// Detected object information.
    /// </summary>
    [MessagePackObject]
    public class DetectedObjectInformation
    {
        /// <summary>
        /// Class id of detected object.
        /// </summary>
        [Key("class_id")]
        public int ClassId { get; set; }

        /// <summary>
        /// Score of detected object.
        /// </summary>
        [Key("score")]
        public float Score { get; set; }

        /// <summary>
        /// Detected object area.
        /// </summary>
        [Key("box")]
        public RectangleRegionParameter Box { get; set; } =
            new RectangleRegionParameter();

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return string.Format(
                "{0}({1}, {2}, {3})",
                this.GetType().FullName, this.ClassId, this.Score, this.Box);
        }
    }
}
